### PR TITLE
kubelet: add new configuration to warn only if some kernel sysctls is not expected.

### LINKED ISSUE #74151: Kubelet sysctl settings cannot be overridden
**What happened**:
The kubelet overrides user defined sysctl parameters silently.

**What you expected to happen**:
Values defined by the administrator in `/etc/sysctl.d/*.conf` files should take priority or at least not be silently overridden.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a file in `/etc/sysctl.d` that conflicts with one that kubelet sets on startup. - `echo "vm.panic_on_oom = 1" > /etc/sysctl.d/30-panic-on-oom.conf` is a good example.
2. Reboot
3. After boot the value specified through sysctl is overridden by kubelet.

**Anything else we need to know?**:
This behavior is particularly tricky to debug because the kubelet is almost always going to start after the systemd-sysctl service. The end result here is that the system boots and claims to have successfully applied all user defined kernel configuration, but then inspecting manually with the `sysctl` utility doesn't show the expected values.

I can understand why the kubelet would want to discover the state of the oom panic toggle on startup, but the changes made for #15091 appear to have completely hard-coded this behavior and actively fight against configuration managed through traditional Linux management options (these parameters would all be valid on the GRUB command line or through systemd setting sysctl parameters on startup). 

I can see a few options that would be a reasonable way to handle this:
1. Offer a configuration option on the kubelet to override the behavior.
2. Stop managing sysctl parameters that are not fatal to the kubelet running and ship recommended parameters as sysctl.d config files through post-install tooling for the kubelet package in a package manager (deb/rpm) or by letting provisioning tools create those files. Parameters that are considered best-practice but are not required for the kubelet to start can throw a warning.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.12.4
- Cloud provider or hardware configuration: Azure
- OS (e.g. from /etc/os-release): Ubuntu 16.04
- Kernel (e.g. `uname -a`): `4.15.0-1037-azure`
- Install tools: AKS Engine

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
The current workaround without this PR is to change sysctls after the kubelet starts each time.
To see use cases, there are https://github.com/kubernetes/kubernetes/issues/66693 https://github.com/kubernetes/kubernetes/issues/90829 https://github.com/kubernetes/kubernetes/issues/50110. 

#### Which issue(s) this PR fixes:
Fixes #74151

#### Special notes for your reviewer:
If ProtectKernelDefaults is set to true, this option will be ignored.

`KernelTunableWarn` is there and cannot be used for current configurations. This PR helps user to use it.
https://github.com/kubernetes/kubernetes/blob/8ce3a2bbef94cada7ca93ca7b1d01ffb68452152/pkg/kubelet/cm/container_manager_linux.go#L426-L444

An issue that may be related: https://github.com/kubernetes/kubeadm/issues/2822 

#### Does this PR introduce a user-facing change?
```release-note
kubelet: add `warningInvalidKernelDefaults` option not to override kernel sysctl default for conflicts.
```

### CODE REVIEW

File: pkg/kubelet/apis/config/types.go [hunk 1]
```diff
@@ -318,6 +318,10 @@ type KubeletConfiguration struct {
 	// flags are not as it expects. Otherwise the Kubelet will attempt to modify
 	// kernel flags to match its expectation.
 	ProtectKernelDefaults bool
+	// warningInvalidKernelDefaults, if ProtectKernelDefaults is set to true, this flag will be ignored.
+	// It only takes effects if ProtectKernelDefaults is false or not set.
+	// If true, kubelet will not override sysctls and only logs warning for those sysctls.
+	WarningInvalidKernelDefaults bool
```
Comment (derekwaynecarr) [inline pkg/kubelet/apis/config/types.go:324]: Can we update the description to make it clear that there is a risk to running with unexpected kernel defaults and usage of this field is appropriate for test/triage environments?  Usage of this field should not open the community to indirectly supporting a mixed set of kernel sysctl settings.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
I am fine with exposing the flag, but we need a more pro-active statement on the ramifications of its usage.  See my suggestion, and let me know if it makes sense?  Thanks!

**Comment by pacoxu** [comment]
> I am fine with exposing the flag, but we need a more pro-active statement on the ramifications of its usage.  See my suggestion, and let me know if it makes sense?  Thanks!

@derekwaynecarr It makes sense and the pr is updated. Do you have time to take a look again before code freezes?

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go [hunk 2]
```diff
@@ -547,6 +547,14 @@ type KubeletConfiguration struct {
 	// Default: false
 	// +optional
 	ProtectKernelDefaults bool `json:"protectKernelDefaults,omitempty"`
+	// warningInvalidKernelDefaults, if ProtectKernelDefaults is set to true, this flag will be ignored.
+	// It only takes effects if ProtectKernelDefaults is false or not set.
```
Comment (cici37) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:551]: nit: we encourage to use field name reference as shown in API doc. `ProtectKernelDefaults` -> `protectKernelDefaults`

Comment (pacoxu) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go]: Fixed. Thanks

**Comment by pacoxu** [comment]
A new feature request in https://github.com/kubernetes/kubernetes/issues/117846.

**Comment by pacoxu** [comment]
ACK @derekwaynecarr PTAL

**Pull request closed without merging**